src/granite/associations.cr

```diff
@@ -1,75 +1,135 @@
 module Granite::Associations
   macro belongs_to(model, **options)
-    {% if model.is_a? TypeDeclaration %}
-      {% method_name = model.var %}
-      {% class_name = model.type %}
-    {% else %}
-      {% method_name = model.id %}
-      {% class_name = options[:class_name] || model.id.camelcase %}
-    {% end %}
+    {%
+      nilable = false
+      if model.is_a? TypeDeclaration
+        name = model.var.id
+        type = model.type
+      else
+        name = model.id
+        type = options[:class_name] || name.camelcase
+      end
+      if type.is_a? Union
+        nilable = type.types.any?(&.resolve?.== Nil)
+        type = type.types.find(&.resolve?.!= Nil)
+      end
+      if nilable
+        nilable_method = name
+        not_nil_method = name + "!"
+      else
+        nilable_method = name + "?"
+        not_nil_method = name
+      end
```

Comment:
I am finding this very confusing.  

If I declare the type as nilable (using a question mark), then I get the default `name` as nilable and an extra getter `name!` as not nilable.

If I declare the type as not nilable (without a question mark), then I get the default option `name` that is not nilable and a `name?` that is nilable.

Is this what you would expect?  

Maybe we always include a `?` option as nilable and a '!' as not nilable for both cases.  The default would be based on the declaration, but you could override the default by using these options.  wdyt?

Reply:
This design came from `field` and `field!` in #181, and they were merged in #346.
Associations should follow the similar syntaxes.
I think `getter!` for nilable type and `getter?` for not nil type are just redundant.

Reply:
I see,  I agree they are redundant, but easy to remember.  I guess if you get an error, you will remember that you declared it one way of the other.

Reply:
Related to #370.  I'm fine with this for the mean time as it fits with the current setup.  However, would like to reach a decision on that PR, which makes things more crystal like id argue versus having to do `field!` for not nilable.